src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -81,31 +85,45 @@ def from_json_dict(data: dict) -> Coordinate:
             group=data["group"],
             artifact=data["artifact"],
             version=data["version"],
+            url=data.get("url", None),
             packaging=data.get("packaging", "jar"),
         )
 
     def to_json_dict(self) -> dict:
-        return {
+        ret = {
             "group": self.group,
             "artifact": self.artifact,
             "version": self.version,
             "packaging": self.packaging,
         }
+        if self.url:
+            ret["url"] = self.url
+        return ret
 
     @classmethod
     def from_coord_str(cls, s: str) -> Coordinate:
+        # if there's a `url=` part, capture the URL after it (otherwise a no-op):
+        s, _, quoted_url = s.partition(",url=")
+        url = url_unquote(quoted_url)  # no effect if `quoted_url` is empty
+
         parts = s.split(":")
         return cls(
             group=parts[0],
             artifact=parts[1],
             version=parts[2],
             packaging=parts[3] if len(parts) == 4 else "jar",
+            url=url if url else None,
         )
 
     def to_coord_str(self, versioned: bool = True) -> str:
+        unversioned = f"{self.group}:{self.artifact}"
+        version_suffix = ""
+        url_suffix = ""
         if versioned:
-            return f"{self.group}:{self.artifact}:{self.version}"
-        return f"{self.group}:{self.artifact}"
+            version_suffix = f":{self.version}"
+        if self.url:
+            url_suffix = f",url={url_quote_plus(self.url)}"
+        return f"{unversioned}{version_suffix}{url_suffix}"
 
     @staticmethod
     def from_jvm_artifact_target(target: Target) -> Coordinate:
```

Comment:
It looks like this is duped with `CoursierResolveSubsystem._coordinate_from_target`.